Update JS Learn Events page (#10914)
mdn/content: files/en-us/learn/javascript/building_blocks/events/index.md

- A video is played, paused, or finishes.
- An error occurs.

You can gather from this (and from glancing at the MDN [event reference](/en-US/docs/Web/Events)) that there are **a lot** of events that can be responded to.

> To respond to an event, you attach an **event handler** to it. This is a block of code (usually a JavaScript function that you as a programmer create) that runs when the event fires.

Reviewer: ```suggestion
To react to an event, you attach an **event handler** to it. This is a block of code (usually a JavaScript function that you as a programmer create) that runs when the event fires.
```
(nit)

  document.body.style.backgroundColor = rndCol;
});
```

> **Note:** You can find the [full source code](https://github.com/mdn/learning-area/blob/master/javascript/building-blocks/events/random-color-addeventlistener.html) for this example on GitHub (also [see it running live](https://mdn.github.io/learning-area/javascript/building-blocks/events/random-color-addeventlistener.html)).

Inside the `addEventListener()` function, we specify two parameters: the name of the event we want to register this handler for, and the code that comprises the handler function we want to run in response to it.

It is fine to make the handler function a separate named function, like this:

```js
const btn = document.querySelector('button');

function random(number) {
  return Math.floor(Math.random() * (number+1));
}

function changeBackground() {
  const rndCol = `rgb(${random(255)}, ${random(255)}, ${random(255)})`;
  document.body.style.backgroundColor = rndCol;
}

btn.addEventListener('click', changeBackground);
```

### Listening for other events

There are many different events that can be fired by a button element. Let's experiment.

> First, make a local copy of [random-color-eventhandlerproperty.html](https://github.com/mdn/learning-area/blob/master/javascript/building-blocks/events/random-color-eventhandlerproperty.html), and open it in your browser.

Reviewer 2: ```suggestion
First, make a local copy of [random-color-addeventlistener.html](https://mdn.github.io/learning-area/javascript/building-blocks/events/random-color-addeventlistener.html), and open it in your browser.
```

```html
<button onclick="alert('Hello, this is my old-fashioned event handler!');">Press me</button>
```

You can find HTML attribute equivalents for many of the event handler properties; however, you shouldn't use these — they are considered bad practice.
It might seem easy to use an event handler attribute if you are doing something really quick, but they quickly become unmanageable and inefficient.

For a start, it is not a good idea to mix up your HTML and your JavaScript, as it becomes hard to read. Keeping your JavaScript separate is a good practice, and if it is in a separate file you can apply it to multiple HTML documents.

Even in a single file, inline event handlers are not a good idea.
One button is OK, but what if you had 100 buttons? You'd have to add 100 attributes to the file; it would quickly turn into a maintenance nightmare.
With JavaScript, you could easily add an event handler function to all the buttons on the page no matter how many there were, using something like this:

```js
const buttons = document.querySelectorAll('button');

for (const button of buttons) {
  button.addEventListener('click', bgChange);
}
```

> > **Note:** Separating your programming logic from your content also makes your site more friendly to search engines.

Reviewer 2: Not sure if this is a thing learners are concerned with at stage, but another point would be that CSP often also disallows the inline JS code.

Reviewer: without diving into CSP, maybe a mention that there are servers where it be blocked from executing, as an extra motivation?

```

You can find HTML attribute equivalents for many of the event handler properties; however, you shouldn't use these — they are considered bad practice.
It might seem easy to use an event handler attribute if you are doing something really quick, but they quickly become unmanageable and inefficient.

For a start, it is not a good idea to mix up your HTML and your JavaScript, as it becomes hard to read. Keeping your JavaScript separate is a good practice, and if it is in a separate file you can apply it to multiple HTML documents.

Even in a single file, inline event handlers are not a good idea.
One button is OK, but what if you had 100 buttons? You'd have to add 100 attributes to the file; it would quickly turn into a maintenance nightmare.
With JavaScript, you could easily add an event handler function to all the buttons on the page no matter how many there were, using something like this:

```js
const buttons = document.querySelectorAll('button');

for (const button of buttons) {
  button.addEventListener('click', bgChange);
}
```

Finally, many common server configurations will disallow inline JavaScript, as a security measure.

> ## Which mechanism should I use?

Reviewer 3: @wbamberg This is all an excellent improvement - in particular making `addEventListener` first and foremost.

FWIW I would remove this section, integrating the key point under the heading  "Other event listener mechanisms" and in each of the parts. 

So in "Other event listener mech" - neither of these methods are recommended as the addEventListener approach is easier and safer to use. Then in the onXX section explain that the limitation that it can only be used once.

I know this is a preference, but I think it is better to introduce the alternatives as options you should not use rather than peers and then explaining later that they aren't really.

Author: Yeah, I agree with this - that section was really a remnant of the previous organization. I just pushed https://github.com/mdn/content/pull/10914/commits/5d4d29c963188ccadb56d10798c0ac4d311e77e1 which removes it.

Reviewer 3: Good fix. I won't add an approval while others are still doing detail review, but the whole thing is much better (and it already wasn't "bad").

> **Note:** See [useful-eventtarget.html](https://github.com/mdn/learning-area/blob/master/javascript/building-blocks/events/useful-eventtarget.html) for the full source code; also see it [running live](https://mdn.github.io/learning-area/javascript/building-blocks/events/useful-eventtarget.html) here.

### Extra properties of event objects

Most event objects have a standard set of properties and methods available on the event object; see the {{domxref("Event")}} object reference for a full list.

> Some event objects add extra properties that are relevant to that particular type of event. For example, the {{domxref("Document/keydown_event", "keydown")}} event fires when the user presses a key, and its event object has a `key` property that tells you which key was pressed:

Reviewer 2: mention here that `key` comes from https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent which is a specialized `Event` object?

> **Note:** for the full source code, see [preventdefault-validation.html](https://github.com/mdn/learning-area/blob/master/javascript/building-blocks/events/preventdefault-validation.html) (also see it [running live](https://mdn.github.io/learning-area/javascript/building-blocks/events/preventdefault-validation.html) here.)

## Event bubbling and capture

> Event bubbling and capture are terms that describe how the browser handles events when elements are nested inside each other.

Reviewer: ```suggestion
Event bubbling and capture are terms that describe phases in how the browser handles events targeted at nested elements.
```
(except for the root, elements are pretty much always nested, so "when elements are nested" isn't very informative)
